File: app/src/main/java/com/example/myittaroostockinventorymanger/ui/item_name/ConfirmDialogFragment.java
package com.example.myittaroostockinventorymanger.ui.item_name;

import android.app.AlertDialog;
import android.app.Dialog;
import android.content.Context;
import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Button;
import android.widget.Toast;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.fragment.app.DialogFragment;
import androidx.lifecycle.ViewModelProvider;

import com.example.myittaroostockinventorymanger.R;
import com.example.myittaroostockinventorymanger.data.entities.Stock;

public class ConfirmDialogFragment extends DialogFragment {


    Button btnYes;

    Button btnNo;

    private AlertDialog alertDialog;
    private Context context;
    private Stock stock;

    private ConfirmDialogFragment() {

    }

    @NonNull
    @Override
    public Dialog onCreateDialog(@Nullable Bundle savedInstanceState) {
        LayoutInflater layoutInflater = getLayoutInflater();
        View view = layoutInflater.inflate(R.layout.confirm_dialog_fragment, null, false);

        btnYes = view.findViewById(R.id.btn_yes);
        btnNo = view.findViewById(R.id.btn_no);

        alertDialog = new AlertDialog.Builder(getContext())
                .setView(view)
                .create();

        alertDialog.getWindow()
                .setBackgroundDrawableResource(R.drawable.rounded_rectangle_white);

        Bundle bundle = getArguments();

        if (bundle != null) {
            stock = bundle.getParcelable(StockNameFragment.EXTRA_DELETE);
        }

        return alertDialog;

    }

    @Nullable
    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {

        ConfirmDialogFragmentViewModel confirmDialogFragmentViewModel = new ViewModelProvider(requireActivity())
                .get(ConfirmDialogFragmentViewModel.class);

        //to delete stock
        btnYes.setOnClickListener(v -> {
            confirmDialogFragmentViewModel.deleteStock(stock);
            alertDialog.cancel();
        });

        btnNo.setOnClickListener(v -> {
            alertDialog.cancel();
        });

        //to show message
        confirmDialogFragmentViewModel.getMessage()
                .observe(getParentFragment().getViewLifecycleOwner(), m -> {

                    String message = m.getContentIfNotHandle();

                    if (message != null) {
                        Toast.makeText(context, message, Toast.LENGTH_SHORT).show();
                    }
                });

        return super.onCreateView(inflater, container, savedInstanceState);
    }

    public static ConfirmDialogFragment getNewInstance(String key, Stock stock) {
        ConfirmDialogFragment confirmDialogFragment = new ConfirmDialogFragment();
        Bundle bundle = new Bundle();
        bundle.putParcelable(key, stock);
        confirmDialogFragment.setArguments(bundle);
        return confirmDialogFragment;
    }

    //context for toast
    @Override
    public void onAttach(@NonNull Context context) {
        super.onAttach(context);
        this.context = context;
    }
}

File: app/src/main/java/com/example/myittaroostockinventorymanger/ui/item_name/ConfirmDialogFragment.kt
package com.example.myittaroostockinventorymanger.ui.item_name

import android.app.AlertDialog
import android.app.Dialog
import android.content.Context
import android.os.Bundle
import android.util.Log
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.Toast
import androidx.fragment.app.DialogFragment
import androidx.fragment.app.viewModels
import androidx.lifecycle.ViewModelProvider
import androidx.lifecycle.switchMap
import androidx.navigation.fragment.navArgs
import androidx.recyclerview.widget.LinearLayoutManager
import com.bumptech.glide.Glide
import com.example.myittaroostockinventorymanger.Application
import com.example.myittaroostockinventorymanger.R
import com.example.myittaroostockinventorymanger.databinding.ConfirmDialogFragmentBinding
import com.example.myittaroostockinventorymanger.event.Event
import com.example.myittaroostockinventorymanger.util.ImageShower
import com.squareup.picasso.Picasso

class ConfirmDialogFragment : DialogFragment() {

    private lateinit var alertDialog: AlertDialog
    private lateinit var context: Context
    private lateinit var selectedIdList: List<Long>
    private lateinit var binding: ConfirmDialogFragmentBinding
    private val arg: ConfirmDialogFragmentArgs by navArgs()
    private lateinit var existingIdList: List<Long>
    private val viewModel: ConfirmDialogFragmentViewModel by viewModels()
    private var itemIdsRelatedToBatch: List<Long> = listOf()
    private lateinit var adapter: WarningRecycleViewAdapter

    override fun onCreateDialog(savedInstanceState: Bundle?): Dialog {
        val layoutInflater = layoutInflater
        binding = ConfirmDialogFragmentBinding.inflate(layoutInflater, null, false)
        val view = binding.root

        val context = getContext()

        if (context != null) {
            this.context = context
        }

        selectedIdList = arg.selectedIdList.toList()

        alertDialog = AlertDialog.Builder(getContext())
            .setView(view)
            .create()
        alertDialog.window
            ?.setBackgroundDrawableResource(R.drawable.rounded_rectangle_white)
        return alertDialog
    }

    override fun onCreateView(
        inflater: LayoutInflater,
        container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View? {

        hideView()

        //set up recyView
        setUpRecycleView(context)

        viewModel.getAllExistItemIdFromBatch()
            .observe(this) {
                existingIdList = it
                viewModel.checkIsItemRelatedToBatch(selectedIdList, it)
            }

        //check for item have related batch data if have delete all data
        viewModel.getIsItemRelatedToBatch
            .observe(this) {
                if (it) {
                    viewModel.deleteBatchesByItemIds(itemIdsRelatedToBatch)
                    viewModel.deleteItems(selectedIdList)
                } else {
                    viewModel.deleteItems(selectedIdList)
                }
                alertDialog.cancel()
            }

        //get item's related data show Warning
        viewModel.getItemIdsRelatedToBatch.switchMap {
            itemIdsRelatedToBatch = it
            viewModel.getBatchWithItemByItemIds(it)
        }.observe(this) {
            adapter.insertData(it)
            binding.txtBody.text = getText(R.string.warning_message)
            showView()
        }

        //set num of delete item
        viewModel.numOfDeleteItems(selectedIdList.size)

        viewModel.getNumOfDeleteItem()
            .observe(this) {
                binding.txtTitle.text = it
            }

        //to delete stock
        binding.btnYes.setOnClickListener { v: View ->
            viewModel.checkIsValidDelete(selectedIdList, existingIdList)
        }

        binding.btnNo.setOnClickListener { v: View -> alertDialog.cancel() }


        //to show message
        viewModel.getMessage()
            .observe(this) { m: Event<String> ->
                val message = m.contentIfNotHandle
                if (message != null) {
                    Toast.makeText(context, message, Toast.LENGTH_SHORT).show()
                }
            }
        return super.onCreateView(inflater, container, savedInstanceState)
    }

    private fun showView() {
        binding.recyWarning.visibility = View.VISIBLE
        binding.txtWarningTitle.visibility = View.VISIBLE
    }

    private fun hideView() {
        binding.recyWarning.visibility = View.GONE
        binding.txtWarningTitle.visibility = View.GONE
    }

    private fun setUpRecycleView(context: Context) {
        adapter = WarningRecycleViewAdapter()

        binding.recyWarning.apply {
            layoutManager = LinearLayoutManager(context)
            adapter = this@ConfirmDialogFragment.adapter
        }
    }
}